src/reduced_order/pod_basis.cpp

```diff
@@ -113,32 +134,46 @@ bool POD<dim>::getPODBasisFromSnapshots() {
         */
         pcout << "Computing POD basis using the method of snapshots..." << std::endl;
 
+        // Get mass matrix
         const bool compute_dRdW = true;
         this->dg->assemble_residual(compute_dRdW);
+        massMatrix.reinit(dg->global_mass_matrix.m(), dg->global_mass_matrix.n());
+        massMatrix.copy_from(dg->global_mass_matrix);
+
+        // Get mass weighted solution snapshots: massWeightedSolutionSnapshots = solutionSnapshots^T * massMatrix * solutionSnapshots
+        dealii::LAPACKFullMatrix<double> tmp(solutionSnapshots.n(), solutionSnapshots.m());
+        massWeightedSolutionSnapshots.reinit(solutionSnapshots.n(), solutionSnapshots.n());
+        solutionSnapshots.Tmmult(tmp, massMatrix);
+        tmp.mmult(massWeightedSolutionSnapshots, solutionSnapshots);
+
+        /* IMPORTANT: This copy of massWeightedSolutionSnapshots is necessary due to an apparent bug in dealii. When
+         * compute_svd() is called on the LAPACKFullMatrix, later trying to convert this LAPACKFullMatrix to a FullMatrix
+         * (necessary in SensitivityPOD) will not give the right matrix. Keep a copy and convert the copy to a FullMatrix
+        */
+        massWeightedSolutionSnapshotsCopy = massWeightedSolutionSnapshots;
 
-        dealii::LAPACKFullMatrix<double> system_matrix(dg->global_mass_matrix.m(), dg->global_mass_matrix.n());
-        system_matrix.copy_from(dg->global_mass_matrix);
-
-        dealii::LAPACKFullMatrix<double> tmp1(snapshot_matrix.n(), snapshot_matrix.m());
-        dealii::LAPACKFullMatrix<double> A(snapshot_matrix.n(), snapshot_matrix.n());
-        snapshot_matrix.Tmmult(tmp1, system_matrix);
-        tmp1.mmult(A, snapshot_matrix);
-
-        A.compute_svd();
+        // Compute SVD of mass weighted solution snapshots: massWeightedSolutionSnapshots = U * Sigma * V^T
+        massWeightedSolutionSnapshots.compute_svd();
 
-        dealii::LAPACKFullMatrix<double> V = A.get_svd_vt();
-        dealii::LAPACKFullMatrix<double> sigma(snapshot_matrix.n(), snapshot_matrix.n());
+        // Get eigenvalues
+        dealii::LAPACKFullMatrix<double> V = massWeightedSolutionSnapshots.get_svd_vt();
+        dealii::LAPACKFullMatrix<double> eigenvectors_T = this->massWeightedSolutionSnapshots.get_svd_vt();
+        eigenvectors.reinit(this->massWeightedSolutionSnapshots.get_svd_vt().n(), this->massWeightedSolutionSnapshots.get_svd_vt().m());
+        eigenvectors_T.transpose(eigenvectors);
 
-        //Form diagonal matrix of singular values
-        for (unsigned int idx = 0; idx < snapshot_matrix.n(); idx++) {
-            sigma(idx, idx) = 1 / A.singular_value(idx);
+        //Form diagonal matrix of inverse singular values
+        eigenvaluesSqrtInverse.reinit(solutionSnapshots.n(), solutionSnapshots.n());
+        for (unsigned int idx = 0; idx < solutionSnapshots.n(); idx++) {
+            eigenvaluesSqrtInverse(idx, idx) = 1 / sqrt(massWeightedSolutionSnapshots.singular_value(idx));
```

Comment:
Which sqrt() does it use? The one in namespace std?

Reply:
I have replaced it with std::sqrt() to make it clearer.